htlcswitch/circuit_map_test.go

```diff
@@ -0,0 +1,403 @@
+	if ks.OutKey == htlcswitch.EmptyCircuitKey {
+		return nil
+	}
+
+	// Open the circuit, implicitly creates a keystone on disk.
+	err = cm.OpenCircuits(ks)
+	if err != nil {
+		return fmt.Errorf("failed to open circuits: %v", err)
+	}
+
+	return nil
+}
+
+// assertKeystoneDeleted checks that a given keystone is deleted from the
+// circuit map.
+func assertKeystoneDeleted(t *testing.T,
+	cm htlcswitch.CircuitLookup, ks htlcswitch.Keystone) {
+
+	c := cm.LookupCircuit(ks.InKey)
+	require.Nil(t, c, "no circuit should be found using InKey")
+
+	if ks.OutKey != htlcswitch.EmptyCircuitKey {
+		c = cm.LookupOpenCircuit(ks.OutKey)
+		require.Nil(t, c, "no circuit should be found using OutKey")
+	}
+}
+
+// assertKeystoneDeleted checks that a given keystone is not deleted from the
+// circuit map.
+func assertKeystoneNotDeleted(t *testing.T,
+	cm htlcswitch.CircuitLookup, ks htlcswitch.Keystone) {
+
+	c := cm.LookupCircuit(ks.InKey)
+	require.NotNil(t, c, "expecting circuit found using InKey")
+
+	if ks.OutKey != htlcswitch.EmptyCircuitKey {
+		c = cm.LookupOpenCircuit(ks.OutKey)
+		require.NotNil(t, c, "expecting circuit found using OutKey")
+	}
+}
+
+// createTestCloseChannelSummery creates a CloseChannelSummery for testing.
+func createTestCloseChannelSummery(tx kvdb.RwTx, isPending bool,
+	chanID lnwire.ShortChannelID) error {
+
+	closedChanBucket, err := tx.CreateTopLevelBucket(closedChannelBucket)
+	if err != nil {
+		return err
+	}
+	outputPoint := wire.OutPoint{Hash: hash1, Index: rand.Uint32()}
+
+	ccs := &channeldb.ChannelCloseSummary{
+		ChanPoint:      outputPoint,
+		ShortChanID:    chanID,
+		ChainHash:      hash1,
+		ClosingTXID:    hash2,
+		CloseHeight:    100,
+		RemotePub:      testEphemeralKey,
+		Capacity:       btcutil.Amount(10000),
+		SettledBalance: btcutil.Amount(50000),
+		CloseType:      channeldb.RemoteForceClose,
+		IsPending:      isPending,
+	}
+	var b bytes.Buffer
+	if err := serializeChannelCloseSummary(&b, ccs); err != nil {
+		return err
+	}
+
+	var chanPointBuf bytes.Buffer
+	if err := writeOutpoint(&chanPointBuf, &outputPoint); err != nil {
+		return err
+	}
+
+	return closedChanBucket.Put(chanPointBuf.Bytes(), b.Bytes())
+}
+
+func serializeChannelCloseSummary(
+	w io.Writer,
+	cs *channeldb.ChannelCloseSummary) error {
+
+	err := channeldb.WriteElements(
+		w,
+		cs.ChanPoint, cs.ShortChanID, cs.ChainHash, cs.ClosingTXID,
+		cs.CloseHeight, cs.RemotePub, cs.Capacity, cs.SettledBalance,
+		cs.TimeLockedBalance, cs.CloseType, cs.IsPending,
+	)
+	if err != nil {
+		return err
+	}
+
+	// If this is a close channel summary created before the addition of
+	// the new fields, then we can exit here.
+	if cs.RemoteCurrentRevocation == nil {
+		return channeldb.WriteElements(w, false)
+	}
+
+	return nil
+}
+
+// writeOutpoint writes an outpoint to the passed writer using the minimal
+// amount of bytes possible.
+func writeOutpoint(w io.Writer, o *wire.OutPoint) error {
```

Comment:
nit: could we just use the function to write to a buffer from `lnwire` instead?

Reply:
good thought! I think there are some works left in terms of putting this circuit map db operations into `channeldb` package and, there are overlaps of writers/encoders between `lnwire` and `channeldb`, which we could fix later.